API transfer demo

libby opened the pull request:
Added ability to optionally store data submitted with a Program. The data can be retrieved via the `RequestId`: `CommandInputQuery RequestId`.
Added endpoint `/api/juno/v1/transfer` for specified JSON:
`{ "payload": {
    "transfer":
      "acctFrom" : "000"
      "acctTo" : "000"
      "amount" : 230.00"
      "currency" : "USD"  
    , "data":
        { "mtmessage": "{data to save}"}
    }
}`

libbylite committed:
Adding Endpoint for transfer.

libbylite committed:
Adding ability to persist data with a Program / Command.

`Program` now takes input text and the term/code to run.
The input is store and can be queried via the RequestId
`CommandInputQuery RequestId`.

sirlensalot commented:
Build is actually broken (test file), so that needs fixing

sirlensalot commented on src/Apps/Juno/ApiDemoHandler.hs:
```diff
@@ -0,0 +1,108 @@
+  where
+    toTx to' from' amount = "transfer(" ++ to' ++ "->" ++ from' ++ "," ++ show (toRational amount) ++ ")"
+
+-- demo API transfer
```
Enclose comment with `{- ... -}` instead of "--", so that a user can copy-paste

sirlensalot commented on src/Apps/Juno/ApiDemoHandler.hs:
```diff
@@ -0,0 +1,108 @@
+--   }
+--}
+--
+data TransferJson = TransferJson { acctFromTrans :: String
```
use more standard naming ... generally we've been using underscores (in case we want to lens up); and a prefix mangler at the beginning (instead of `Trans` at the end) is standard. Suggestion: _tAcctFrom, _tAcctTo, _tAmount, _tCurrency

sirlensalot commented on src/Apps/Juno/ApiDemoHandler.hs:
```diff
@@ -0,0 +1,108 @@
+  toJSON (TransferDataJson databody) =
+      object ["mtmessage" .= databody]
+
+instance FromJSON TransferDataJson where
```
use `withObject` instead of casing

sirlensalot commented on src/Apps/Juno/ApiDemoHandler.hs:
```diff
@@ -0,0 +1,108 @@
+             ]
+
+instance FromJSON TransferJson where
+  parseJSON (Object v) = TransferJson <$> v .: "acctFrom"
```
use `withObject` instead of casing

sirlensalot commented on src/Apps/Juno/Command.hs:
```diff
@@ -19,17 +19,22 @@ import Data.Aeson (encode, object, (.=))
 import Data.Aeson.Encode.Pretty (encodePretty)
 
 import Juno.Types (CommandEntry(..), CommandResult(..))
+import Juno.Types.Base (RequestId)
```
Not really a fan of qualified imports from project-local files unless necessary.

sirlensalot commented on src/Apps/Juno/Command.hs:
```diff
@@ -19,17 +19,22 @@ import Data.Aeson (encode, object, (.=))
 
 -- state of hopper
-newtype JunoEnv = JunoEnv {getStateMVar :: MV.MVar (DEval.PersistentState, Map.Map DEval.TransactionId SWIFT)}
+newtype JunoEnv = JunoEnv {getStateMVar :: MV.MVar (DEval.PersistentState
```
arrgh really hating our use of tuples inside of an mvar. Should really be a type, with lenses to ease the garbage code in `runCommand`

sirlensalot commented on src/Apps/Juno/Command.hs:
```diff
@@ -38,80 +43,85 @@ setBalances :: Map.Map Text Rational -> DEval.PersistentState -> DEval.Persisten
+runCommand env (cmd', rid) = do
   mvar <- return $ getStateMVar env
-  (ps, ss) <- MV.takeMVar mvar -- persistent s, swift
+  (ps, ss, store) <- MV.takeMVar mvar -- persistent s, swift, prog input store
```
yuck. at a minimum capture the whole tuple for our no-op re-puts. ie, `org@(ps,ss,store)` ...

sirlensalot commented on src/Apps/Juno/Command.hs:
```diff
@@ -38,80 +43,85 @@ setBalances :: Map.Map Text Rational -> DEval.PersistentState -> DEval.Persisten
   case readHopper $ unCommandEntry cmd' of
     Left err -> do
-      MV.putMVar mvar (ps,ss)
+      MV.putMVar mvar (ps, ss, store)
```
... so that here we can just say `org` instead of this. better yet, add a `let` for a `noop` function that does this, and use everywhere we're copypasting this noop.

sirlensalot commented on src/Apps/Juno/Parser.hs:
```diff
@@ -29,28 +30,34 @@ import Juno.Hoplite.Term (HopliteTerm(..))
 
 data HopperLiteAdminCommand
-  = Program HopliteTerm
+  = Program Text HopliteTerm -- input to store with term.
```
Ah, this is the code that broke the test.

sirlensalot commented on src/Apps/Juno/ApiDemoHandler.hs:
```diff
@@ -0,0 +1,108 @@
+                                               <*> ((v .: "payload") >>= (.: "data"))
+    parseJSON _ = mzero
+
+goodTrans = JSON.encode $ TransferJson "000" "001" 24192.10 "USD"
```
putting underscores before un-exported stuff gets rid of warnings, like the test values here.

sirlensalot commented on src/Apps/Juno/Command.hs:
```diff
@@ -38,80 +43,85 @@ setBalances :: Map.Map Text Rational -> DEval.PersistentState -> DEval.Persisten
                             Right (DEval.InterpreterOutput (DEval.InterpreterDiff _balsDiff ops)
                                                         nextPs) -> do
-                                MV.putMVar mvar (nextPs, ss)
+                                store' <- return $ Map.insert rid input store
```
inline store' into the mvar put (avoid unnecc. vars, esp all of these 'tick ones, wherever poss)

sirlensalot commented on src/Juno/Consensus/Commit.hs:
```diff
@@ -67,7 +67,7 @@ applyCommand :: Monad m => UTCTime -> Command -> Raft m (NodeID, CommandResponse
   apply <- view (rs.applyLogEntry)
   logApplyLatency cmd
-  result <- apply _cmdEntry
+  result <- apply (_cmdEntry, _cmdRequestId)
```
hmmm... should we just be handing off the entire Command? the incremental add of "requestId" seems to suggest we might want other stuff in the future. What's more it took me a while to figure out what changed to add it to `applyFun`, `runCommand` etc, and is kind of "spec specific" now. Just dumping off the whole command seems more future-proof.

libby replied:
You're right, that is more future proof. I'll switch to passing the entire command in.

sirlensalot commented:
OK review done. Nice work on the parsers in the API stuff btw, I didn't get a chance to look at it after you did that. Muuucho mejor :)

libbylite committed:
Cleaning up and fixing tests after Stu's PR review.

libby commented:
Thanks for the feedback!

libbylite committed:
Use `Command` instead of `(CommandEntry, RequestId)` in runCommand.

Re: PR suggestion, it is more future proof and cleaner to pass
Command around with contains fields needed for `runCommand`.

libbylite committed:
Use `withObject` instead of pattern matching in aeson tc instances.
